redis/conn.go

```diff
@@ -238,10 +257,26 @@ func DialContext(ctx context.Context, network, address string, options ...DialOp
 		}
 
 		tlsConn := tls.Client(netConn, tlsConfig)
-		if err := tlsConn.Handshake(); err != nil {
-			netConn.Close()
+
+		errc := make(chan error, 2)
+		var timer *time.Timer // for cancelling TLS handshake
+		if d := do.tlsHandshakeTimeout; d != 0 {
+			timer = time.AfterFunc(d, func() {
+				errc <- tlsHandshakeTimeoutError{}
+			})
+		}
+		go func() {
+			err := tlsConn.Handshake()
+			if timer != nil {
+				timer.Stop()
+			}
+			errc <- err
```

Comment:
Could be cleaned up a little by deferring the timer stop (should be nearly zero overhead for this since 1.14) and sending the handshake result straight down the channel.
```suggestion
			if timer != nil {
				defer timer.Stop()
			}
			errc <- tlsConn.Handshake()
```